modules/core/include/opencv2/core/cvdef.h

```diff
@@ -330,6 +330,142 @@ Cv64suf;
+}                                                                                                     \
+
+#define __CV_ENUM_FLAGS_BITWISE_AND_EQ(EnumType, Arg1Type)                                            \
+static inline EnumType& operator&=(EnumType& _this, const Arg1Type& val)                              \
+{                                                                                                     \
+    _this = static_cast<EnumType>(static_cast<int>(_this) & static_cast<int>(val));                   \
+    return _this;                                                                                     \
+}                                                                                                     \
+
+#define __CV_ENUM_FLAGS_BITWISE_XOR_EQ(EnumType, Arg1Type)                                            \
+static inline EnumType& operator^=(EnumType& _this, const Arg1Type& val)                              \
+{                                                                                                     \
+    _this = static_cast<EnumType>(static_cast<int>(_this) ^ static_cast<int>(val));                   \
+    return _this;                                                                                     \
+}                                                                                                     \
+
+#define CV_ENUM_CLASS_EXPOSE(EnumType, ...)                                                           \
+__CV_EXPAND(__CV_CAT(__CV_ENUM_CLASS_EXPOSE_, __CV_VA_NUM_ARGS(__VA_ARGS__))(EnumType, __VA_ARGS__)); \
+
+#define CV_ENUM_FLAGS(EnumType)                                                                       \
+__CV_ENUM_FLAGS_LOGICAL_NOT      (EnumType);                                                          \
+__CV_ENUM_FLAGS_LOGICAL_EQ       (EnumType, int);                                                     \
+__CV_ENUM_FLAGS_LOGICAL_NOT_EQ   (EnumType, int);                                                     \
+                                                                                                      \
+__CV_ENUM_FLAGS_BITWISE_NOT      (EnumType);                                                          \
+__CV_ENUM_FLAGS_BITWISE_OR       (EnumType, EnumType, EnumType);                                      \
+__CV_ENUM_FLAGS_BITWISE_AND      (EnumType, EnumType, EnumType);                                      \
+__CV_ENUM_FLAGS_BITWISE_XOR      (EnumType, EnumType, EnumType);                                      \
+                                                                                                      \
+__CV_ENUM_FLAGS_BITWISE_OR_EQ    (EnumType, EnumType);                                                \
+__CV_ENUM_FLAGS_BITWISE_AND_EQ   (EnumType, EnumType);                                                \
+__CV_ENUM_FLAGS_BITWISE_XOR_EQ   (EnumType, EnumType);                                                \
```

Comment:
I am observing multiple warnings with GCC 7:
```
../opencv/modules/core/include/opencv2/core/cvdef.h:460:44: warning: extra ‘;’ [-Wpedantic]
__CV_ENUM_FLAGS_BITWISE_NOT      (EnumType);                                                          \ 
                                            ^                                                                                      
../opencv/modules/core/include/opencv2/core/mat.hpp:566:1: note: in expansion of macro ‘CV_ENUM_FLAGS’
```
Probably we have to remove last semicolon in this and other declarations.

Reply:
Thank you! I noticed that from the qt image warnings..
- If we remove the semicolons from the call, the editor would make wrong indentation (calls are made in several other places)
- If we let the semicolons here, we will have warnings when `-Wpedantic` is enabled

Therefore, the best solution is to add a semicolon to the macro header itself. I already made #12631 to do that.